File: src/Array/Longest_subarray_with_sum_K.java
package Array;

public class Longest_subarray_with_sum_K {
    public static void main(String[] args) {
        int[] arr = {4, 1, 1, 1, 2, 3, 5};
        int k = 5;
        int result = longestSubarrayWithSumK(arr, k);
        System.out.println("Length of the longest subarray: " + result);
    }

    private static int longestSubarrayWithSumK(int[] arr, int k) {
        int start = 0, currentSum = 0, maxLength = 0;

        for (int end = 0; end < arr.length; end++) {
            currentSum += arr[end];

            while (currentSum > k && start <= end) {
                currentSum -= arr[start];
                start++;
            }

            if (currentSum == k) {
                maxLength = Math.max(maxLength, end - start + 1);
            }
        }
        return maxLength;
    }
}

File: Array/longestSubarrayWithSumK.ts
// Brute Force - 

const longestSubarrayWithSumK = (nums: number[], value:number): number => {
  let length = nums.length;
  let maxLength = 0;

  for(let i = 0; i < length; i++){
    let sum = 0;
    for(let j = i; j < length; j++){
      sum += nums[j];

      if(sum == value){
        maxLength = Math.max(maxLength, j - i + 1);
      }
    }
  }

  return maxLength;
}



// Optimal Solution - 

const longestSubarrayWithSumK = (nums: number[], k:number): number => {
  let prefixSumMap = new Map<number, number>();
    let sum = 0, maxLength = 0;

    for (let i = 0; i < nums.length; i++) {
        sum += nums[i];

        if (sum === k) {
            maxLength = i + 1;
        }

        if (prefixSumMap.has(sum - k)) {
            maxLength = Math.max(maxLength, i - prefixSumMap.get(sum - k)!);
        }

        if (!prefixSumMap.has(sum)) {
            prefixSumMap.set(sum, i);
        }
    }

    return maxLength;
}


console.log(longestSubarrayWithSumK([10, 5, 2, 7, 1, -10], 15));
